scripts/codesys-plc-info.nse

```diff
@@ -0,0 +1,127 @@
+]]
+
+---
+-- @usage
+-- nmap --script codesys-plc-info
+--
+-- @output
+-- 1740/udp open|filtered encore
+-- | codesys-plc-info: 
+-- |   192.168.20.12: 
+-- |     deviceAddress: 192.168.20.12
+-- |     targetVendor: WAGO
+-- |     targetName: WAGO 750-8215 PFC200 G2 4ETH CAN USB
+-- |     deviceName: PFC200V3-4538EF
+-- |     targetID: 0x1006120b
+-- |     targetType: 0x1000
+-- |_    targetVersion: 5.15.4.0
+--
+-- @args codesys-plc-info.timeout timespec defining how long to wait for a
+--       response. (default 3s)
+
+--
+-- Version 0.1
+-- Created 23/06/2021 - v0.1 - created by Andreas Galauner <agalauner@rapid7.com>
+--
+
+author = "Andreas Galauner"
+license = "Same as Nmap--See https://nmap.org/book/man-legal.html"
+categories = {"default", "discovery", "safe"}
+portrule = shortport.portnumber({1740,1741,1742,1743}, "udp")
+
+--- Returns the network interface used to send packets to a target host.
+-- @param target host to which the interface is used.
+-- @return interface Network interface used for target host.
+local getInterface = function(target)
+  -- First, create dummy UDP connection to get interface
+  local sock = nmap.new_socket()
+  local status, err = sock:connect(target, "12345", "udp")
+  if not status then
+    stdnse.verbose1("%s", err)
+    return
+  end
+  local status, address, _, _, _ = sock:get_info()
+  if not status then
+    stdnse.verbose1("%s", err)
+    return
+  end
+  for _, interface in pairs(nmap.list_interfaces()) do
+    if interface.address == address then
+      return interface
+    end
+  end
+end
+
+local function fail (err) return stdnse.format_output(false, err) end
+
+action = function(host, port)
+  local timeout = stdnse.parse_timespec(stdnse.get_script_args(SCRIPT_NAME .. ".timeout"))
+  timeout = (timeout or 3) * 1000
+
+  -- Try to determine the interface we can reach our target with. We need this for the name service request
+  local iface = getInterface(host)
+  if not iface then
+    return fail(string.format("Couldn't get interface for target IP address %s", host))
+  end
+  
+  local socket = nmap.new_socket("udp")
+  socket:set_timeout(timeout)
+
+  -- Bind to source port 1740, because we are using port index 0 in the name service request
+  -- We need to send the packet from this source port, otherwise the PLC doesn't seem to reply
+  local status, err = socket:bind(nil, 1740)
+  if not status then
+    return fail(string.format("Bind failed: %s", err))
+  end
+
+  -- Generate the name service request to send and send it out
+  local cs = codesys3.CodesysV3.NameServiceRequest:new(0, iface.address, iface.netmask)
+  local packet = tostring(cs)
+
+  local status, err = socket:sendto(host, port, packet)
+  if not status then
+    return fail(string.format("Sendto failed: %s", err))
+  end
+
+  -- Receive the responses from the PLCs and parse them
+  local result = {}
+  repeat
+    local data
+    status, data = socket:receive()
```

Comment:
EDIT: You can pretty much ignore this comment, since I had apparently completely forgotten about connected UDP sockets (see note above on sendto) which resolve the issue. However, I don't have the heart to delete it now that it's written, since I spent a good amount of time researching it.
___
Since the source address is hard-coded here, you may end up receiving responses from hosts other than the one this script is running against. My tests show that there is no issue with multiple UDP sockets binding to the same source port and sending data, though TCP would fail the `connect()` call immediately if the source port was already in use. So you need to introduce some other way of ensuring you don't contend for the same source port between multiple scripts running concurrently. A mutex would work, using the address and port number as the "key" like so:

```lua
local mutex = nmap.mutex( ("%s:%d"):format(iface.address, 1740) )
mutex "lock"
-- socket send/receive ops
mutex "done"
```

You can reduce chance of contention by switching up the source port among the 4 allowable values. Something like this would probably work:

```lua
local portindex = 0
local locked = false
-- first try to find an unused port
local mutex
for i=0, 3 do
  portindex = i
  mutex = nmap.mutex( ("%s:%d"):format(iface.address, 1740 + portindex) )
  locked = mutex "trylock"
  if locked then
    break
  end
end
-- If that didn't work, pick one at random and block until it's available
if not locked then
  portindex = math.random(0,3)
  mutex = nmap.mutex( ("%s:%d"):format(iface.address, 1740 + portindex) )
  mutex "lock"
end
-- socket ops, etc.
mutex "done"
```